crates/runtime/src/dataaccelerator/metadata/duckdb.rs

```diff
@@ -0,0 +1,116 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{path::Path, sync::Arc};
+
+use data_components::duckdb::DuckDB;
+use duckdb::AccessMode;
+
+use super::AcceleratedMetadataProvider;
+use super::{METADATA_DATASET_COLUMN, METADATA_METADATA_COLUMN, METADATA_TABLE_NAME};
+use crate::component::dataset::{acceleration::Engine, Dataset};
+use crate::dataaccelerator::{get_accelerator_engine, DuckDBAccelerator};
+use db_connection_pool::duckdbpool::DuckDbConnectionPool;
+
+pub struct AcceleratedMetadataDuckDB {
+    pool: Arc<DuckDbConnectionPool>,
+}
+
+impl AcceleratedMetadataDuckDB {
+    pub async fn try_new(
+        dataset: &Dataset,
+        create_if_file_not_exists: bool,
+    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        let accelerator = get_accelerator_engine(Engine::DuckDB)
+            .await
+            .ok_or("DuckDB accelerator engine not available")?;
+        let duckdb_accelerator = accelerator
+            .as_any()
+            .downcast_ref::<DuckDBAccelerator>()
+            .ok_or("Accelerator is not a DuckDBAccelerator")?;
+
+        let duckdb_file = duckdb_accelerator
+            .duckdb_file_path(dataset)
+            .ok_or("Acceleration mode is not file-based.")?;
+        if !create_if_file_not_exists && !Path::new(&duckdb_file).exists() {
+            return Err("DuckDB file does not exist.".into());
+        }
+
+        let pool = DuckDbConnectionPool::new_file(&duckdb_file, &AccessMode::ReadWrite)
+            .map_err(|e| e.to_string())?;
+
+        Ok(Self {
+            pool: Arc::new(pool),
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl AcceleratedMetadataProvider for AcceleratedMetadataDuckDB {
+    async fn get_metadata(&self, dataset: &str) -> Option<String> {
+        let mut db_conn = Arc::clone(&self.pool).connect_sync().ok()?;
+        let duckdb_conn = DuckDB::duckdb_conn(&mut db_conn)
+            .ok()?
+            .get_underlying_conn_mut();
+        let query = format!(
+            "SELECT {METADATA_METADATA_COLUMN} FROM {METADATA_TABLE_NAME} WHERE {METADATA_DATASET_COLUMN} = ?",
+        );
+        let mut stmt = duckdb_conn.prepare(&query).ok()?;
+        let mut rows = stmt.query([dataset]).ok()?;
+
+        let metadata: Option<String> = if let Some(row) = rows.next().ok()? {
+            Some(row.get(0).ok()?)
+        } else {
+            None
+        };
+
+        metadata
+    }
+
+    async fn set_metadata(
+        &self,
+        dataset: &str,
+        metadata: &str,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+        let mut db_conn = Arc::clone(&self.pool)
+            .connect_sync()
+            .map_err(|e| e.to_string())?;
+        let duckdb_conn = DuckDB::duckdb_conn(&mut db_conn)
+            .map_err(|e| e.to_string())?
+            .get_underlying_conn_mut();
+
+        let create_if_not_exists = format!(
```

Comment:
I wonder if this whole schema format should be in `crates/runtime/src/dataaccelerator/metadata.rs`

Reply:
Technically there could be variations in how the table is constructed depending on the underlying provider, which is why I didn't centralize it.